### PR TITLE
Better concurrency

### PR DESCRIPTION
**Pull request by boaz-codota**

### CODE REVIEW

File: src/main/java/com/tabnine/TabNineCompletionContributor.java [hunk 1]
```diff
@@ -32,6 +35,10 @@
 public class TabNineCompletionContributor extends CompletionContributor {
     private final TabNineGateway process = DependencyContainer.singletonOfTabNineGateway();
 
+    private int consecutiveRestarts = 0;
```
Comment (dimacodota) [inline src/main/java/com/tabnine/TabNineCompletionContributor.java:38]: why not AtomicInteger ?

**Review by dimacodota — COMMENTED** [review]
see my questions

File: src/main/java/com/tabnine/TabNineCompletionContributor.java [hunk 2]
```diff
@@ -110,14 +123,28 @@ private AutocompleteResponse retrieveCompletions(CompletionParameters parameters
                 req.regionIncludesEnd = (end == doc.getTextLength());
 
                 try {
-                    return AppExecutorUtil.getAppExecutorService().submit(() -> this.process.request(req))
-                            .get(COMPLETION_TIME_THRESHOLD, TimeUnit.MILLISECONDS);
+                    Future<AutocompleteResponse> request = AppExecutorUtil.getAppExecutorService().submit(() -> this.process.request(req));
+
+                    activeRequests.add(request);
```
[the remark this answers is not in the document]
Comment (boaz-codota) [reply src/main/java/com/tabnine/TabNineCompletionContributor.java:128]: hmm could be cleaned after the request was completed. How would you do that? run a background thread that cleans finished tasks?

File: src/main/java/com/tabnine/TabNineCompletionContributor.java (hunk 1 above)
Comment (boaz-codota) [reply src/main/java/com/tabnine/TabNineCompletionContributor.java:38]: No need, it is not multithreaded

Comment (boaz-codota) [reply src/main/java/com/tabnine/TabNineCompletionContributor.java:38]: If it was multithreaded, IntelliJ wasn't hang when the response didn't arrive (that is why we needed to implement timeouts...)

File: src/main/java/com/tabnine/TabNineCompletionContributor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dimacodota) [reply src/main/java/com/tabnine/TabNineCompletionContributor.java:128]: why not remove it after the ` this.process.request(req)` in the submit?

File: src/main/java/com/tabnine/TabNineCompletionContributor.java (hunk 1 above)
Comment (dimacodota) [reply src/main/java/com/tabnine/TabNineCompletionContributor.java:38]: oh, I thought you're modifying it in the submit... nevermind

**Comment by boaz-codota** [comment]
Looks like this:
image

**Review by dimacodota — APPROVED** [review]
LGTM

**Pull request merged**